src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+end
+
+function write_discrete_to_proposal(state::JacobianPassState, addr, value)
+    value
+end
+
+# write_discrete_to_model
+
+macro write_discrete_to_model(addr, value)
+    quote write_discrete_to_model($(esc(inv_state)), $(esc(addr)), $(esc(value))) end
+end
+
+function write_discrete_to_model(state::FirstPassState, addr, value)
+    state.constraints[addr] = value
+    value
+end
+
+function write_discrete_to_model(state::JacobianPassState, addr, value)
+    value
+end
+
+# read continuous from proposal
+
+macro read_continuous_from_proposal(addr)
+    quote read_continuous_from_proposal($(esc(inv_state)), $(esc(addr))) end
+end
+
+function read_continuous_from_proposal(state::FirstPassState, addr)
+    state.u_cont_reads[addr] = state.u[addr]
+    state.u[addr]
+end
+
+function read_continuous_from_proposal(state::JacobianPassState, addr)
+    if haskey(state.u_key_to_index, addr)
+        state.input_arr[state.u_key_to_index[addr]]
+    else
+        state.u[addr]
+    end
+end
+
+# read continuous from model
+
+macro read_continuous_from_model(addr)
+    quote read_continuous_from_model($(esc(inv_state)), $(esc(addr))) end
+end
+
+function read_continuous_from_model(state::FirstPassState, addr)
+    state.t_cont_reads[addr] = state.trace[addr]
+    state.trace[addr]
+end
+
+function read_continuous_from_model(state::JacobianPassState, addr)
+    if haskey(state.t_key_to_index, addr)
+        state.input_arr[state.t_key_to_index[addr]]
+    else
+        state.trace[addr]
+    end
+end
+
+# read continuous from model retained (optional, for efficiency)
+
+macro read_continuous_from_model_retained(addr)
+    quote read_continuous_from_model_retained($(esc(inv_state)), $(esc(addr))) end
+end
+
+function read_continuous_from_model_retained(state::FirstPassState, addr)
+    push!(state.marked_as_retained, addr)
+    state.trace[addr]
+end
+
+function read_continuous_from_model_retained(state::JacobianPassState, addr)
+    state.trace[addr] # read directly from the trace, instead of the array
+end
+
+# write_continuous_to_proposal
+
+macro write_continuous_to_proposal(addr, value)
+    quote write_continuous_to_proposal($(esc(inv_state)), $(esc(addr)), $(esc(value))) end
+end
+
+function write_continuous_to_proposal(state::FirstPassState, addr, value)
+    has_value(state.u_back, addr) && error("Proposal address $addr already written to")
+    state.u_back[addr] = value
+    state.u_cont_writes[addr] = value
+    value
+end
+
+function write_continuous_to_proposal(state::JacobianPassState, addr, value)
+    state.output_arr[state.cont_u_back_key_to_index[addr]] = value
+end
+
+# write_continuous_to_model
+
+macro write_continuous_to_model(addr, value)
+    quote write_continuous_to_model($(esc(inv_state)), $(esc(addr)), $(esc(value))) end
+end
+
+function write_continuous_to_model(state::FirstPassState, addr, value)
+    has_value(state.constraints, addr) && error("Model address $addr already written to")
+    state.constraints[addr] = value
```

Comment:
My understanding is a bit shaky on this, but want to ask: if this `constraints` is the same `constraints` as in the API for `Gen.generate`, then will adding constraints here mess with the score that would be computed by `Gen.get_score`?

Reply:
I think this `constraints` is just `t'` from the TeX doc -- it's where you're keeping track of the things that have been proposed to the model. It will ultimately be passed to `update`.  (I think it shouldn't contain any of the addresses that were originally passed to `generate` when creating an initial MCMC trace, because your proposal shouldn't be changing any of those values.)

Reply:
Oh good.  I think when I originally read that name I thought `Gen.generate` would be used to construct the new trace, and in that context these things would become "constraints" or something like that... was going to suggest changing the name.

Actually... I think I'd still like to suggest that :)  Would a name like `new_choices` make sense?

Reply:
:+1: for `new_choices` :-)

Reply:
Constraints contains a subset of the choices in t' from the Tex doc. It does not contain those elements of t' that are going to be copied implicitly from t.